### DISCUSSION TITLE
Provide a way to access OpenSSL source code via plain HTTP now that ftp.openssl.org is no more

### DISCUSSION THREAD

**Discussion by Googulator** [openssl/openssl#24722] [Q&A]
With http://ftp.openssl.org now gone, all official URLs where OpenSSL's source code can be downloaded from are now HTTPS-only. This creates a chicken-and-egg problem, where OpenSSL is required to access its own source code.

This causes a problem for projects like live-bootstrap (https://github.com/fosslinux/live-bootstrap), which now has to rely on 3rd-party mirrors to download OpenSSL - mirrors that might switch to HTTPS-only at any moment without notice, as has happened several times in the past.

**Comment by nhorman** [comment]
what does livebootstrap currently use to download sources?

**Comment by stikonas** [comment]
> what does livebootstrap currently use to download sources?

We had two options (see https://github.com/fosslinux/live-bootstrap/blob/master/steps/openssl-3.0.13/sources):

```
http://ftp.openssl.org/source/openssl-3.0.13.tar.gz 88525753f79d3bec27d2fa7c66aa0b92b3aa9498dafd93d7cfa4b3780cdae313
http://ftp.openssl.org/source/old/3.0/openssl-3.0.13.tar.gz 88525753f79d3bec27d2fa7c66aa0b92b3aa9498dafd93d7cfa4b3780cdae313
```

**Comment by nhorman** [comment]
I'm sorry, I mean what tool are you using to download your sources?  presumably some minimal set of binaries to start the bootstrapping process.  From the download_distfiles.sh script in the repository, it looks like you require curl at minimum to already be available on the host system to run your process.  So I'm trying to figure out why openssl can't be a requirement on your host system as well.  In fact, it seems like it has to already be a requirement, as you have several other sources files in other steps that require https

**Comment by stikonas** [comment]
Oh there are a couple of ways:
1. Download everything on the host system and store it on some disk. No issue here...
2. Download smaller number of sources (these are the ones you noted to be https, they are less than 300 MiB of tarballs). Basically what is needed to bootstrap Linux (earlier kernels have no networking support anyway) and then newly booted system proceeds downloading remaining sources. Perhaps openssl could be included there too though it will likely increase amount of initial pre-downloaded sources...

**Comment by levitte** [comment]
> > what does livebootstrap currently use to download sources?
> 
> We had two options (see https://github.com/fosslinux/live-bootstrap/blob/master/steps/openssl-3.0.13/sources):
> 
> ```
> http://ftp.openssl.org/source/openssl-3.0.13.tar.gz 88525753f79d3bec27d2fa7c66aa0b92b3aa9498dafd93d7cfa4b3780cdae313
> http://ftp.openssl.org/source/old/3.0/openssl-3.0.13.tar.gz 88525753f79d3bec27d2fa7c66aa0b92b3aa9498dafd93d7cfa4b3780cdae313
> ```

So uhm...  can't you simply change `http:` to `https:`?  Like @nhorman says, https://github.com/fosslinux/live-bootstrap/blob/master/download-distfiles.sh uses `curl`, which can handle HTTPS these days.

I'll point out that https://github.com/fosslinux/live-bootstrap/blob/master/steps/curl-8.5.0/sources already uses HTTPS, isn't that an indication that you bootstrap with a HTTPS enabled `curl`?

**Comment by nhorman** [comment]
And if your minimal boot loader downloads the remainder of the sources, isn't it feasible to just add openssl to that list of minimal sources to get the rest of everything down?  It seems an implied requirement in this system that anything using https be part of the "minimal"  set.  Though if you download openssl as part of that set, your overall set could be reduced, as the minimal ramdisk could then be used to download many more sources that were only available via https

**Comment by Googulator** [comment]
download-distfiles.sh is not used in the normal bootstrap path, but only by the CI infrastructure, to speed up builds (and enable building in a sandboxed environment with no runtime network access). The actual standard bootstrap path uses a special build of curl, built without SSL support, to download openssl and its prerequisites; all of these downloads need to happen over plain HTTP as a result.

One of the goals of the project is to keep the total size of all source code that needs to be predownloaded on the host system (using Python's Requests library) under 256MiB, which is the size of the largest known unmanaged flash memory chip with a simple interface like SPI. All larger flash chips include firmware that could be used to compromise the bootstrap's anti-Trusting-Trust guarantees. We currently meet this goal, but if openssl and all of its dependencies were included in the predownloaded set, the 256MiB goal would be almost, if not outright, impossible to meet.

More generally, it's not great practice to require the very protocol implemented by a library for downloading the source code for that library, even if that protocol is widespread - of course, it's unavoidable for protocols like the basic HTTP, but in this case, it's entirely avoidable. Only serving OpenSSL sources over SSL is comparable to lzip's former practice of only offering lzip-compressed archives of its own source code.

**Comment by nhorman** [comment]
but, its also not good practice to allow downloads without communication security in place..

In fact, I would make the argument that this isn't a problem with the fact that an SSL connection is required to download the openssl sources, its that OpenSSL isn't nominally designed to be used on smaller embedded platforms (which this target arguably is, given its size constraints).

Currently, the typical openssl library DSO  on x86_64 weighs in at around 5MB, and the libssl library at 0.8MB (using the fedora distro build as a guide).  You can certainly shrink that down quite a bit by disabling features you don't need, using SMALL_FOOTPRINT, etc.

The only dependency that either library requires is potentially libz or libbrotli, both of which I believe can be disabled, but at worst the libz weight is 1.3MB

If the addtions of about 7MB is over your capacity for your target, it seems the path forward here is to use a smaller footprint ssl library to preform your downloads for the purpose of bootstrapping.  Something like mbedtls, perhaps.

I'm sorry, but we're not going to enable a non-secure download path to enable easier retrieval of any sources, and I would predict that this is going to be a pervasive theme moving forward.  Given efforts like https://www.eff.org/encrypt-the-web, theres going to be a movement in which the expectation is that more an more downloads will be https-only

**Comment by t8m** [comment]
So you are afraid that your image might be compromised due to the firmware on the flash but not afraid that you download something random over http instead of the true openssl library sources? I assume you aren't verifying the tarball signature either as that would require having gpg and the openssl public key on the image.

**Reply by kroeckx** [reply]
The sha256 is saved in their repository, they know which file to expect. I assume to update the repository they use at least some form of crypto.

**Comment by paulidale** [comment]
Why such resistance here?  It is a quite reasonable request IMO.

Mandating a working TLS installation to download the world's premier TLS implementation is rubbish.

**Reply by nhorman** [reply]
I disagree @paulidale on two points

1) A project as widespread as OpenSSL is under significant scrutiny in how we do our software distribution, and its not worth the risk to continue to offer software without properly securing its transmission.  Even with the sha256 sum present, if we have an http server get mitm-ed and pushes a bogus tarball to a user that doesn't bother to check the sha, the news report won't be "careless user does a careless thing".  It will be "openssl doesn't bother to secure their own website with their own software"

2) bootstrapping always requires some level of pre-existing software.  We may have this conversation in equal measure about curl or wget, as neither can be downloaded without a functional http library.  There's general acceptance of that though because it's just understood that for that particular egg/chicken  combo we assume the existence of the proverbial egg.  In my view we're at a tipping point where functional TLS libraries are assumed.
